lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// +build !solaris,!darwin solaris,cgo darwin,cgo
+
+package fswatcher
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/ignore"
+	"github.com/zillode/notify"
+)
+
+type eventType int
+
+const (
+	nonRemove eventType = 1
+	remove              = 2
+	mixed               = 3
+)
+
+// Not meant to be changed, but must be changeable for tests
+var (
+	maxFiles       = 512
+	maxFilesPerDir = 128
+)
+
+type event struct {
+	path         string
+	firstModTime time.Time
+	lastModTime  time.Time
+	evType       eventType
+}
+
+type eventBatch map[string]*event
```

Comment:
Does it need to be a map to a pointer though? If its not modified out of the map, then not using pointers would cause less gc pressure, and save a deref. Yet I am quite bad at recognizing when to use one but not the other.